{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "a4a0703a-437e-4746-9bb1-4ff63cb2b211",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "import json\n",
    "import ast"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efde9568-a2f9-436c-9b0a-48e7ffce1743",
   "metadata": {},
   "source": [
    "The libraries below are for display purposes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "id": "906de2d8-4d8c-4911-95dd-e6b32bad0cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from IPython.display import display, Markdown\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "id": "abe2df00-2741-435c-90c4-cf692872a998",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apiQuery(tier, query, variables):\n",
    "    if tier == 'prod':\n",
    "        url = 'https://hub.datacommons.cancer.gov/api/graphql'\n",
    "        token = os.environ['PRODAPI']\n",
    "    elif tier == 'stage':\n",
    "        #Note that use of Stage is for example purposes only, actual submissions should use the production URL.  If you wish to run tests on Stage, please contact the helpdesk.\n",
    "        url = 'https://hub-stage.datacommons.cancer.gov/api/graphql'\n",
    "        token = os.environ['STAGEAPI']\n",
    "    else:\n",
    "        return('Please provide either \"stage\" or \"prod\" as tier values')\n",
    "    headers = {\"Authorization\": f\"Bearer {token}\"}\n",
    "    try:\n",
    "        if variables is None:\n",
    "            result = requests.post(url = url, headers = headers, json={\"query\": query})\n",
    "        else:\n",
    "            result = requests.post(url = url, headers = headers, json = {\"query\":query, \"variables\":variables})\n",
    "        if result.status_code == 200:\n",
    "            return result.json()\n",
    "        else:\n",
    "            print(f\"Error: {result.status_code}\")\n",
    "            return result.content\n",
    "    except requests.exceptions.HTTPError as e:\n",
    "        return(f\"HTTP Error: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "id": "09fc59f8-1cee-46d3-b14e-b32c248929b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "subid = '9ba3152c-4194-4a30-ba9d-10bbba735302'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 170,
   "id": "fc4093f3-4c06-4739-a888-d71f652ec29a",
   "metadata": {},
   "outputs": [],
   "source": [
    "neweq = \"\"\"\n",
    "query retrieveReleasedDataByID(\n",
    "    $submissionID: String!,\n",
    "    $nodeType: String!\n",
    "    $nodeID: String!\n",
    "){\n",
    "retrieveReleasedDataByID(\n",
    "    submissionID: $submissionID,\n",
    "    nodeType: $nodeType\n",
    "    nodeID: $nodeID\n",
    "){\n",
    "    submissionID\n",
    "    status\n",
    "    dataCommons\n",
    "    dataCommonsDisplayName\n",
    "    studyID\n",
    "    nodeType\n",
    "    nodeID\n",
    "    props\n",
    "}\n",
    "}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "id": "98454fd5-ee0e-4947-a682-a40e8802ca08",
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_nodes_query = \"\"\"\n",
    "query getSubmissionNodes(\n",
    "    $_id: String!,\n",
    "    $nodeType: String!, \n",
    "    $status: String,\n",
    "    $first: Int, \n",
    "    $offset:Int, \n",
    "    $orderBy: String, \n",
    "    $sortDirection:String\n",
    ") {\n",
    "getSubmissionNodes(\n",
    "    submissionID: $_id\n",
    "    nodeType: $nodeType\n",
    "    status: $status\n",
    "    first: $first\n",
    "    offset: $offset\n",
    "    orderBy: $orderBy\n",
    "    sortDirection: $sortDirection\n",
    ") {\n",
    "    total\n",
    "    IDPropName\n",
    "    properties\n",
    "    nodes {\n",
    "        nodeID\n",
    "        nodeType\n",
    "        status\n",
    "        props\n",
    "    }\n",
    "    }\n",
    "}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "id": "9b1a357c-2bd8-481f-bfaf-4e6c951090dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "severity = 'All'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 173,
   "id": "7af4250d-6ac4-4205-b87c-70cb0cb4126a",
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_nodes_vars = {'_id':subid, 'nodeType':nodetype, 'status':severity, 'first':-1, 'offset':0, 'orderBy':'studyID', 'sortDirection':'desc'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "id": "5901c295-6f88-4c90-ad48-8203a483e89f",
   "metadata": {},
   "outputs": [],
   "source": [
    "subnodes_res = apiQuery('stage', submission_nodes_query, submission_nodes_vars)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "id": "fe3b3e80-fa58-41d4-8dcc-67cef69aa5ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "columns = [\"nodeID\",\"nodeType\",\"sample_id\",\"sample_type\",\"sample_description\",\"sample_type_category\",\"sample_tumor_status\",\"sample_anatomic_site\",\"sample_age_at_collection\",\"derived_from_specimen\",\"biosample_accession\",\"participant.study_participant_id\"]\n",
    "subnodes_df = pd.DataFrame(columns=columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "id": "88181d84-4af4-4bfb-899e-3177fa4beb30",
   "metadata": {},
   "outputs": [],
   "source": [
    "for node in subnodes_res['data']['getSubmissionNodes']['nodes']:\n",
    "    #pprint.pprint(f\"{subnodes_res['data']['getSubmissionNodes']}\\n\")\n",
    "    temp = json.loads(node['props'])\n",
    "    temp['nodeID'] = node['nodeID']\n",
    "    temp['nodeType'] = node['nodeType']\n",
    "    subnodes_df.loc[len(subnodes_df)] = temp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 177,
   "id": "072e5f06-e0c8-4ca2-ad27-ebfd0893ffa8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#display(Markdown(subnodes_df.to_markdown()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 190,
   "id": "faca4f60-382f-4eb4-88f1-b9fb9b0c7cf3",
   "metadata": {},
   "outputs": [],
   "source": [
    "difflist = []\n",
    "for index, row in subnodes_df.iterrows():\n",
    "    nodetype = row['nodeType']\n",
    "    nodeID = row['nodeID']\n",
    "    variables = {'submissionID': subid , 'nodeType': nodetype, 'nodeID': nodeID, 'first': -1}\n",
    "    diffres = apiQuery('stage', neweq, variables)\n",
    "    dfcollection = {}\n",
    "    #difflist = []\n",
    "    for entry in diffres['data']['retrieveReleasedDataByID']:\n",
    "        propstuff = ast.literal_eval(entry['props'])\n",
    "        temp_df = pd.DataFrame(propstuff, index=[entry['submissionID']])\n",
    "        dfcollection[entry['submissionID']] = temp_df\n",
    "        keylist = list(dfcollection.keys())\n",
    "        if len(keylist) >= 2:\n",
    "            df1 = dfcollection[keylist[0]]\n",
    "            df2 = dfcollection[keylist[1]]\n",
    "            diff_df = pd.concat([df1, df2]).drop_duplicates(keep=False)\n",
    "            difflist.append(diff_df)\n",
    "            #display(Markdown(diff_df.to_markdown()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "id": "5ceac68a-d345-4c41-abf9-f32bafb7ab3e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "|                                      | reference_genome_assembly   | design_description                                                                                                                                                                                                                                                                                                                                                                          | library_id   | library_strategy   | library_source_material   | library_source_molecule   | library_selection   | platform   | instrument_model      | genomic_info_id                           |   bases |   number_of_reads |   avg_read_length |   coverage | library_layout   | sequence_alignment_software                      |\n",
       "|:-------------------------------------|:----------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------|:-------------------|:--------------------------|:--------------------------|:--------------------|:-----------|:----------------------|:------------------------------------------|--------:|------------------:|------------------:|-----------:|:-----------------|:-------------------------------------------------|\n",
       "| 9ba3152c-4194-4a30-ba9d-10bbba735302 | GRCh38                      | WXS Hybrid Selection paired-end                                                                                                                                                                                                                                                                                                                                                             | lib1         | ChIP-Seq           | Bulk Tissue               | Viral RNA                 | MSLL                | Illumina   | Illumina NovaSeq 6000 | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_1 |     648 |               116 |             65.22 |          1 | Paired-End       | Nucleotide Sequence Alignment Software Name Text |\n",
       "| 1d75dfe1-f4e9-4445-9836-e11f55d92761 | GRCh38                      | WXS Hybrid Selection paired-end                                                                                                                                                                                                                                                                                                                                                             | nan          | ChIP-Seq           | Bulk Tissue               | Viral RNA                 | MSLL                | Illumina   | Illumina NovaSeq 6000 | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_1 |     648 |               116 |             65.22 |          1 | Paired-End       | nan                                              |\n",
       "| 9ba3152c-4194-4a30-ba9d-10bbba735302 | GRCh37                      | Anchored Multiplex PCR (AMP(TM)) chemistry is purpose-built to accurately identify both simple and complex genetic mutations by leveraging the power of bidirectional primers and molecular barcodes, or MBCs, from low nucleic acid input in tissue or blood. AMP chemistry is also flexible, so it can be used for applications in DNA, RNA, and ctDNA sequencing across most tumor type. | lib2         | CTS                | Bulk Tissue               | Viral RNA                 | MSLL                | BGISEQ     | Illumina MiSeq        | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_2 |     879 |               530 |             41.91 |          2 | Paired-End       | test1                                            |\n",
       "| 1d75dfe1-f4e9-4445-9836-e11f55d92761 | GRCh37                      | Anchored Multiplex PCR (AMP(TM)) chemistry is purpose-built to accurately identify both simple and complex genetic mutations by leveraging the power of bidirectional primers and molecular barcodes, or MBCs, from low nucleic acid input in tissue or blood. AMP chemistry is also flexible, so it can be used for applications in DNA, RNA, and ctDNA sequencing across most tumor type. | nan          | CTS                | Bulk Tissue               | Viral RNA                 | MSLL                | BGISEQ     | Illumina MiSeq        | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_2 |     879 |               530 |             41.91 |          2 | Paired-End       | nan                                              |\n",
       "| 9ba3152c-4194-4a30-ba9d-10bbba735302 | GRCh38                      | WXS Hybrid Selection paired-end                                                                                                                                                                                                                                                                                                                                                             | lib3         | CLONEEND           | Bulk Tissue               | Transcriptome             | MSLL                | LS 454     | Illumina NovaSeq 6000 | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_3 |     127 |               527 |            151.42 |          3 | Paired-End       | test2                                            |\n",
       "| 1d75dfe1-f4e9-4445-9836-e11f55d92761 | GRCh38                      | WXS Hybrid Selection paired-end                                                                                                                                                                                                                                                                                                                                                             | nan          | CLONEEND           | Bulk Tissue               | Transcriptome             | MSLL                | LS 454     | Illumina NovaSeq 6000 | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_3 |     127 |               527 |            151.42 |          3 | Paired-End       | nan                                              |\n",
       "| 9ba3152c-4194-4a30-ba9d-10bbba735302 | GRCh37                      | Anchored Multiplex PCR (AMP(TM)) chemistry is purpose-built to accurately identify both simple and complex genetic mutations by leveraging the power of bidirectional primers and molecular barcodes, or MBCs, from low nucleic acid input in tissue or blood. AMP chemistry is also flexible, so it can be used for applications in DNA, RNA, and ctDNA sequencing across most tumor type. | lib4         | snATAC-Seq         | Bulk Tissue               | Transcriptome             | MSLL                | Ultima     | Illumina MiSeq        | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_4 |     762 |               943 |            148.27 |          4 | Paired-End       | seq3                                             |\n",
       "| 1d75dfe1-f4e9-4445-9836-e11f55d92761 | GRCh37                      | Anchored Multiplex PCR (AMP(TM)) chemistry is purpose-built to accurately identify both simple and complex genetic mutations by leveraging the power of bidirectional primers and molecular barcodes, or MBCs, from low nucleic acid input in tissue or blood. AMP chemistry is also flexible, so it can be used for applications in DNA, RNA, and ctDNA sequencing across most tumor type. | nan          | snATAC-Seq         | Bulk Tissue               | Transcriptome             | MSLL                | Ultima     | Illumina MiSeq        | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_4 |     762 |               943 |            148.27 |          4 | Paired-End       | nan                                              |\n",
       "| 9ba3152c-4194-4a30-ba9d-10bbba735302 | GRCh37                      | Anchored Multiplex PCR (AMP(TM)) chemistry is purpose-built to accurately identify both simple and complex genetic mutations by leveraging the power of bidirectional primers and molecular barcodes, or MBCs, from low nucleic acid input in tissue or blood. AMP chemistry is also flexible, so it can be used for applications in DNA, RNA, and ctDNA sequencing across most tumor type. | lib5         | FINISHING          | Bulk Tissue               | Transcriptome             | Other               | Illumina   | Illumina MiSeq        | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_5 |     668 |               843 |            403.05 |          5 | Single-indexed   | seq4                                             |\n",
       "| 1d75dfe1-f4e9-4445-9836-e11f55d92761 | GRCh37                      | Anchored Multiplex PCR (AMP(TM)) chemistry is purpose-built to accurately identify both simple and complex genetic mutations by leveraging the power of bidirectional primers and molecular barcodes, or MBCs, from low nucleic acid input in tissue or blood. AMP chemistry is also flexible, so it can be used for applications in DNA, RNA, and ctDNA sequencing across most tumor type. | nan          | FINISHING          | Bulk Tissue               | Transcriptome             | Other               | Illumina   | Illumina MiSeq        | dg.4DFC/840c7e9c-8900-4d1d-96667ce99c1e_5 |     668 |               843 |            403.05 |          5 | Single-indexed   | nan                                              |"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "report_df = pd.concat(difflist)\n",
    "display(Markdown(report_df.to_markdown()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 192,
   "id": "6dda2be2-618b-45c3-96a9-d1750434634e",
   "metadata": {},
   "outputs": [],
   "source": [
    "report_df.to_csv('/home/pihl/temp/warningdiffs.csv', sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16fdbdef-0466-4ad7-83b3-c50d15fe5c4b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
