{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "a4a0703a-437e-4746-9bb1-4ff63cb2b211",
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import os\n",
    "import json\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "efde9568-a2f9-436c-9b0a-48e7ffce1743",
   "metadata": {},
   "source": [
    "The libraries below are for display purposes only"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "906de2d8-4d8c-4911-95dd-e6b32bad0cf1",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import display, Markdown\n",
    "import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "abe2df00-2741-435c-90c4-cf692872a998",
   "metadata": {},
   "outputs": [],
   "source": [
    "def apiQuery(tier, query, variables):\n",
    "    if tier == 'prod':\n",
    "        url = 'https://hub.datacommons.cancer.gov/api/graphql'\n",
    "        token = os.environ['PRODAPI']\n",
    "    elif tier == 'stage':\n",
    "        #Note that use of Stage is for example purposes only, actual submissions should use the production URL.  If you wish to run tests on Stage, please contact the helpdesk.\n",
    "        url = 'https://hub-stage.datacommons.cancer.gov/api/graphql'\n",
    "        token = os.environ['STAGEAPI']\n",
    "    else:\n",
    "        return('Please provide either \"stage\" or \"prod\" as tier values')\n",
    "    headers = {\"Authorization\": f\"Bearer {token}\"}\n",
    "    try:\n",
    "        if variables is None:\n",
    "            result = requests.post(url = url, headers = headers, json={\"query\": query})\n",
    "        else:\n",
    "            result = requests.post(url = url, headers = headers, json = {\"query\":query, \"variables\":variables})\n",
    "        if result.status_code == 200:\n",
    "            return result.json()\n",
    "        else:\n",
    "            print(f\"Error: {result.status_code}\")\n",
    "            return result.content\n",
    "    except requests.exceptions.HTTPError as e:\n",
    "        return(f\"HTTP Error: {e}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2926b885-7f77-4561-a517-ad44702efd2f",
   "metadata": {},
   "source": [
    "Queries needed to get the list of errors"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "fc4093f3-4c06-4739-a888-d71f652ec29a",
   "metadata": {},
   "outputs": [],
   "source": [
    "error_query = \"\"\"\n",
    "query retrieveReleasedDataByID(\n",
    "    $submissionID: String!,\n",
    "    $nodeType: String!\n",
    "    $nodeID: String!\n",
    "){\n",
    "retrieveReleasedDataByID(\n",
    "    submissionID: $submissionID,\n",
    "    nodeType: $nodeType\n",
    "    nodeID: $nodeID\n",
    "){\n",
    "    submissionID\n",
    "    status\n",
    "    dataCommons\n",
    "    dataCommonsDisplayName\n",
    "    studyID\n",
    "    nodeType\n",
    "    nodeID\n",
    "    props\n",
    "}\n",
    "}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "98454fd5-ee0e-4947-a682-a40e8802ca08",
   "metadata": {},
   "outputs": [],
   "source": [
    "submission_nodes_query = \"\"\"\n",
    "query getSubmissionNodes(\n",
    "    $_id: String!,\n",
    "    $nodeType: String!, \n",
    "    $status: String,\n",
    "    $first: Int, \n",
    "    $offset:Int, \n",
    "    $orderBy: String, \n",
    "    $sortDirection:String\n",
    ") {\n",
    "getSubmissionNodes(\n",
    "    submissionID: $_id\n",
    "    nodeType: $nodeType\n",
    "    status: $status\n",
    "    first: $first\n",
    "    offset: $offset\n",
    "    orderBy: $orderBy\n",
    "    sortDirection: $sortDirection\n",
    ") {\n",
    "    total\n",
    "    IDPropName\n",
    "    properties\n",
    "    nodes {\n",
    "        nodeID\n",
    "        nodeType\n",
    "        status\n",
    "        props\n",
    "    }\n",
    "    }\n",
    "}\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "7aaac297-fd63-469b-9617-074d416fda3e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def diffDataFrame(subid, nodetype, nodeID, tier, query):\n",
    "    difflist = []\n",
    "    variables = {'submissionID': subid , 'nodeType': nodetype, 'nodeID': nodeID}\n",
    "    diffres = apiQuery(tier, query, variables)\n",
    "    dfcollection = {}\n",
    "    if 'errors' in diffres:\n",
    "        return None\n",
    "    else:\n",
    "        for entry in diffres['data']['retrieveReleasedDataByID']:\n",
    "            #propstuff = ast.literal_eval(entry['props'])\n",
    "            propstuff = json.loads(entry['props'])\n",
    "            temp_df = pd.DataFrame(propstuff, index=[entry['submissionID']])\n",
    "            dfcollection[entry['submissionID']] = temp_df\n",
    "            keylist = list(dfcollection.keys())\n",
    "            if len(keylist) >= 2:\n",
    "                df1 = dfcollection[keylist[0]]\n",
    "                df2 = dfcollection[keylist[1]]\n",
    "                diff_df = pd.concat([df1, df2]).drop_duplicates(keep=False)\n",
    "                difflist.append(diff_df)\n",
    "        report_df = pd.concat(difflist)\n",
    "        return report_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac1c414f-8196-4db0-8405-7ebb79d3271f",
   "metadata": {},
   "source": [
    "The cell below contains the variables that need to be set by the user:\n",
    "- subid: The submission ID to check (can be obtained in the graphical interface)\n",
    "- severity: Should be set to 'All'\n",
    "- nodelist: the node names containing the warnings that are to be checked\n",
    "- outputdirectory: Where the summary spreadsheets should be saved.  Must have '/' at the end\n",
    "- tier - Must be either 'prod' or 'stage'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "9b1a357c-2bd8-481f-bfaf-4e6c951090dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "subid = 'c0ceb350-9318-4a48-875a-dd956a73e211'\n",
    "severity = 'All'\n",
    "nodelist = ['genomic_info', 'image', 'proteomic']\n",
    "outputdirectory = '/media/sf_VMShare/WarningSummary/'\n",
    "tier = 'stage'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "3e41d5ed-c517-4e86-bc0e-20f8270b968e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/markdown": [
       "| submission_id                        | genomic_info_id   | library_id                             |   bases |   number_of_reads |   avg_read_length |   coverage | reference_genome_assembly   | custom_assembly_fasta_file_for_alignment    | design_description                     | library_strategy   | library_layout   | library_selection   | platform   | instrument_model         | sequence_alignment_software                  | reporter_label        | methylation_platform                        | library_source_material   | library_source_molecule   | crdc_id                                      |\n",
       "|:-------------------------------------|:------------------|:---------------------------------------|--------:|------------------:|------------------:|-----------:|:----------------------------|:--------------------------------------------|:---------------------------------------|:-------------------|:-----------------|:--------------------|:-----------|:-------------------------|:---------------------------------------------|:----------------------|:--------------------------------------------|:--------------------------|:--------------------------|:---------------------------------------------|\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-GENO-115990    | Gallia_est_omnis_divisa_in_partes_tres |     720 |               863 |            846.52 |          8 | Not specified in data       | scientia_et_labor                           | damnant_quod_non_intellegunt           | Spatial-tx         | Paired-End       | Unspecified         | BGISEQ     | 454 GS FLX+              | e_causa_ignota                               | e_causa_ignota        | scientiae_et_patriae                        | Bulk Cells                | Genomic DNA               | nascentes_morimur_finisque_ab_origine_pendet |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-GENO-115990    | Gallia_est_omnis_divisa_in_partes_tres |     720 |               863 |            846.52 |          8 | Not specified in data       | scientia_et_labor                           | Illumina NextSeq 2500                  | Spatial-tx         | Paired-End       | Unspecified         | BGISEQ     | 454 GS FLX+              | e_causa_ignota                               | e_causa_ignota        | scientiae_et_patriae                        | Bulk Cells                | Genomic DNA               | nascentes_morimur_finisque_ab_origine_pendet |\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-GENO-185213    | scientia_et_labor                      |     552 |               620 |            808.98 |          0 | GRCh37                      | Macte_animo_Generose_puer_sic_itur_ad_astra | e_causa_ignota                         | CLONE              | Single-indexed   | PCR                 | Ultima     | Illumina NextSeq 500     | scientia_imperii_decus_et_tutamen            | a_bene_placito        | a_bene_placito                              | Single-nuclei             | Not Reported              | barba_crescit_caput_nescit                   |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-GENO-185213    | scientia_et_labor                      |     552 |               620 |            808.98 |          0 | GRCh37                      | Macte_animo_Generose_puer_sic_itur_ad_astra | Illumina NextSeq 2500                  | CLONE              | Single-indexed   | PCR                 | Ultima     | Illumina NextSeq 500     | scientia_imperii_decus_et_tutamen            | a_bene_placito        | a_bene_placito                              | Single-nuclei             | Not Reported              | barba_crescit_caput_nescit                   |\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-GENO-645746    | scientia_imperii_decus_et_tutamen      |     525 |               253 |            469.13 |          3 | GRCh38                      | saltus_in_demonstrando                      | cacatum_non_est_pictum                 | scRNA-Seq          | Single-indexed   | ChIP                | Helicos    | PacBio Sequel II         | nascentes_morimur_finisque_ab_origine_pendet | sapiens_qui_prospicit | Macte_animo_Generose_puer_sic_itur_ad_astra | Not Reported              | Transcriptome             | cacatum_non_est_pictum                       |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-GENO-645746    | scientia_imperii_decus_et_tutamen      |     525 |               253 |            469.13 |          3 | GRCh38                      | saltus_in_demonstrando                      | Illumina NextSeq 2500                  | scRNA-Seq          | Single-indexed   | ChIP                | Helicos    | PacBio Sequel II         | nascentes_morimur_finisque_ab_origine_pendet | sapiens_qui_prospicit | Macte_animo_Generose_puer_sic_itur_ad_astra | Not Reported              | Transcriptome             | cacatum_non_est_pictum                       |\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-GENO-808685    | cacatum_non_est_pictum                 |     961 |               492 |            948.83 |          1 | GRCh37-lite                 | labor_ipse_voluptas                         | Gallia_est_omnis_divisa_in_partes_tres | FL-cDNA            | Single-indexed   | Unspecified         | Helicos    | AB 3130 Genetic Analyzer | id_quod_plerumque_accidit                    | scientia_et_labor     | scientia,_aere_perennius                    | Single-nuclei             | Not Reported              | a_bene_placito                               |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-GENO-808685    | cacatum_non_est_pictum                 |     961 |               492 |            948.83 |          1 | GRCh37-lite                 | labor_ipse_voluptas                         | Illumina NextSeq 2500                  | FL-cDNA            | Single-indexed   | Unspecified         | Helicos    | AB 3130 Genetic Analyzer | id_quod_plerumque_accidit                    | scientia_et_labor     | scientia,_aere_perennius                    | Single-nuclei             | Not Reported              | a_bene_placito                               |"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node: genomic_info\n",
      "Diff report: None\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/markdown": [
       "| submission_id                        | study_link_id   | de_identification_method_type   | de_identification_method_description         | de_identification_software   | license   | citation_or_DOI                   | species        | image_modality   | imaging_equipment_manufacturer   | imaging_equipment_model   | imaging_software                       | imaging_protocol         | organ_or_tissue                         | performed_imaging_study_typeCode       | longitudinal_temporal_event_type   |   longitudinal_temporal_event_offset | crdc_id              |\n",
       "|:-------------------------------------|:----------------|:--------------------------------|:---------------------------------------------|:-----------------------------|:----------|:----------------------------------|:---------------|:-----------------|:---------------------------------|:--------------------------|:---------------------------------------|:-------------------------|:----------------------------------------|:---------------------------------------|:-----------------------------------|-------------------------------------:|:---------------------|\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-IMAGE-666518 | Automatic                       | pace_tua                                     | imperium_in_imperio          | CC BY 3.0 | nanos_gigantum_humeris_insidentes | a_bene_placito | GM               | a_bene_placito                   | scientiae_cedit_mare      | Gallia_est_omnis_divisa_in_partes_tres | scientia,_aere_perennius | Sigmoid colon                           | Gallia_est_omnis_divisa_in_partes_tres | Baseline                           |                               904.75 | labor_ipse_voluptas  |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-IMAGE-666518 | Automatic                       | pace_tua                                     | imperium_in_imperio          | CC BY 3.0 | nanos_gigantum_humeris_insidentes | a_bene_placito | GM               | Orbitrap                         | Orbitrap                  | Gallia_est_omnis_divisa_in_partes_tres | scientia,_aere_perennius | Sigmoid colon                           | Gallia_est_omnis_divisa_in_partes_tres | Baseline                           |                               904.75 | labor_ipse_voluptas  |\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-IMAGE-715541 | Semiautomatic                   | nascentes_morimur_finisque_ab_origine_pendet | id_quod_plerumque_accidit    | CC BY 3.0 | a_bene_placito                    | a_bene_placito | DM               | barba_crescit_caput_nescit       | scientia_et_sapientia     | saltus_in_demonstrando                 | sapiens_qui_prospicit    | Overlapping lesion of ill-defined sites | scientia,_aere_perennius               | Enrollment                         |                               778.1  | scientiae_cedit_mare |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-IMAGE-715541 | Semiautomatic                   | nascentes_morimur_finisque_ab_origine_pendet | id_quod_plerumque_accidit    | CC BY 3.0 | a_bene_placito                    | a_bene_placito | DM               | Orbitrap                         | Orbitrap                  | saltus_in_demonstrando                 | sapiens_qui_prospicit    | Overlapping lesion of ill-defined sites | scientia,_aere_perennius               | Enrollment                         |                               778.1  | scientiae_cedit_mare |"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node: image\n",
      "Diff report: None\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/markdown": [
       "| submission_id                        | proteomic_info_id   | aliquot_id          | analytical_fractions   | instrument_make                 | proteomic_instrument_model   | proteomic_design_description           | manufacturer_model_name   | crdc_id                      |\n",
       "|:-------------------------------------|:--------------------|:--------------------|:-----------------------|:--------------------------------|:-----------------------------|:---------------------------------------|:--------------------------|:-----------------------------|\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-PROTEOMIC-270724 | imperium_in_imperio | Proteome               | faber_est_suae_quisque_fortunae | labor_ipse_voluptas          | haec_olim_meminisse_iuvabit            | Orbitrap                  | saltus_in_demonstrando       |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-PROTEOMIC-270724 | imperium_in_imperio | Proteome               | faber_est_suae_quisque_fortunae | Orbitrap                     | haec_olim_meminisse_iuvabit            | Orbitrap                  | nan                          |\n",
       "| c0ceb350-9318-4a48-875a-dd956a73e211 | GC-PROTEOMIC-393569 | scientia_et_labor   | Phosphoproteome        | damnant_quod_non_intellegunt    | scientia_et_labor            | Gallia_est_omnis_divisa_in_partes_tres | Orbitrap                  | damnant_quod_non_intellegunt |\n",
       "| 807cdd3c-1066-49ad-ae58-95b6a7987744 | GC-PROTEOMIC-393569 | scientia_et_labor   | Phosphoproteome        | damnant_quod_non_intellegunt    | Orbitrap                     | Gallia_est_omnis_divisa_in_partes_tres | Orbitrap                  | nan                          |"
      ],
      "text/plain": [
       "<IPython.core.display.Markdown object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Node: proteomic\n",
      "Diff report: None\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for node in nodelist:\n",
    "    nodedflist = []\n",
    "    node_vars = {'_id':subid, 'nodeType':node, 'status':severity, 'first':-1, 'offset':0, 'orderBy':'studyID', 'sortDirection':'desc'}\n",
    "    nodedata_res = apiQuery(tier, submission_nodes_query, node_vars)\n",
    "    #Set up the dataframe needed to query for errors\n",
    "    for result in nodedata_res['data']['getSubmissionNodes']['nodes']:\n",
    "        nodetype = result['nodeType']\n",
    "        nodeid = result['nodeID']\n",
    "        report_df = diffDataFrame(subid, nodetype, nodeid, 'stage', error_query)\n",
    "        if report_df is not None:\n",
    "            nodedflist.append(report_df)\n",
    "    report_df = pd.concat(nodedflist)\n",
    "    report_df.index.name = 'submission_id'\n",
    "    report_df.to_csv(f\"{outputdirectory}{node}_warning_diffs.csv\", sep=\"\\t\")\n",
    "    #Uncomment the line below if you want to display the output tables in this notebook.\n",
    "    print(f\"Node: {node}\\nDiff report: {display(Markdown(report_df.to_markdown()))}\\n\")\n",
    "        \n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13d2aef3-4ce4-4cfc-9253-9b799134044a",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
